internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,364 @@
+// a facade, and when notified initiates a connection to the controller's SSH server.
+// When the controller's SSH server receives this connection, it initiates its own
+// SSH connection down the tunnel and the sshSessionWorker pipes it to the SSHD
+// on the machine.
+type sshSessionWorker struct {
+	tomb tomb.Tomb
+
+	logger           Logger
+	facadeClient     FacadeClient
+	agent            agent.Agent
+	connectionGetter ConnectionGetter
+}
+
+// NewWorker returns an SSH session worker.
+func NewWorker(cfg WorkerConfig) (worker.Worker, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &sshSessionWorker{
+		logger:           cfg.Logger,
+		facadeClient:     cfg.FacadeClient,
+		agent:            cfg.Agent,
+		connectionGetter: cfg.ConnectionGetter,
+	}
+
+	w.tomb.Go(w.loop)
+	return w, nil
+}
+
+// loop starts the workers main loop. It watches for incoming SSH connection requests
+// and handles them by initiating a reverse SSH connection to the controller and piping
+// it to the local sshd of the machine agent's machine.
+func (w *sshSessionWorker) loop() error {
+	machineId := w.agent.CurrentConfig().Tag().Id()
+	sw, err := w.facadeClient.WatchSSHConnRequest(machineId)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case changes := <-sw.Changes():
+			for _, docID := range changes {
+				err := w.handleConnection(docID)
+				if err != nil {
+					w.logger.Errorf("failed to handle connection %q: %v", docID, err)
+				}
+			}
+		}
+	}
+}
+
+// Kill implements the worker.Worker interface.
+func (w *sshSessionWorker) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait implements the worker.Worker interface.
+func (w *sshSessionWorker) Wait() error {
+	return w.tomb.Wait()
+}
+
+// handleConnection handles initiating a reverse SSH connection to the controller
+// and piping it to the local sshd of the machine agent's machine.
+//
+// This function does the following:
+//  1. Gets the controllers address and ephemeral public key for the connection.
+//  2. Verifies the address is known to this machine agent.
+//  3. Adds the ephemeral public key to the authorized_keys2 file.
+//  4. Dials the controllers SSH server expecting an SSH connection to come
+//     back from the controller's SSH server.
+//  5. Pipes the connection to the local sshd.
+//  6. On connection close, removes the ephemeral public key from the authorized_keys2 file.
+func (w *sshSessionWorker) handleConnection(docID string) error {
+	reqParams, err := w.facadeClient.GetSSHConnRequest(docID)
+	if err != nil {
+		w.logger.Errorf("failed to get SSH connection request: %v", err)
+		return errors.Trace(err)
+	}
+
+	ctrlAddress := reqParams.ControllerAddress.Values()[0]
+	ephemeralPublicKey := string(reqParams.EphemeralPublicKey)
+
+	if err := w.verifyControllerAddress(ctrlAddress); err != nil {
```

Comment:
Maintaining controller api addresses is already done by the `api-address-updater` worker. There's no need to go to the controller to get them and then compare to what's in the agent config file.

Reply:
This is more about checking the address coming in is known before making an SSH connecting back down it.

Reply:
I've actually removed this now.